From: Florian Fainelli <f.fainelli@gmail.com>
Date: Wed, 17 Feb 2016 18:43:22 -0800
Subject: net: dsa: Unregister slave_dev in error path
Patch-mainline: v4.5-rc6
Git-commit: 73dcb556538a4192222c3a919a51e1701bae553b
References: bsc#1026030 FATE#321670

With commit 0071f56e46da ("dsa: Register netdev before phy"), we are now trying
to free a network device that has been previously registered, and in case of
errors, this will make us hit the BUG_ON(dev->reg_state != NETREG_UNREGISTERED)
condition.

Fix this by adding a missing unregister_netdev() before free_netdev().

Fixes: 0071f56e46da ("dsa: Register netdev before phy")
Signed-off-by: Florian Fainelli <f.fainelli@gmail.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Acked-by: Benjamin Poirier <bpoirier@suse.com>
---
 net/dsa/slave.c |    1 +
 1 file changed, 1 insertion(+)

--- a/net/dsa/slave.c
+++ b/net/dsa/slave.c
@@ -1207,6 +1207,7 @@ int dsa_slave_create(struct dsa_switch *
 	ret = dsa_slave_phy_setup(p, slave_dev);
 	if (ret) {
 		netdev_err(master, "error %d setting up slave phy\n", ret);
+		unregister_netdev(slave_dev);
 		free_netdev(slave_dev);
 		return ret;
 	}
